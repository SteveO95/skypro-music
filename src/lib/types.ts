export enum FilterType {
	author = 'исполнителю',
	year = 'году выпуска',
	genre = 'жанру',
}

export type TrackDataType = {
	_id: number
	name: string
	author: string
	release_date: string
	genre: string[]
	duration_in_seconds: number
	album: string
	logo: { type: string; data: unknown[] }
	track_file: string
	staredUser: number[]
}

export type UserDataType = {
	email: string
	username: string
	_id: number
}

export type TokenType = {
	access: string
	refresh: string
}

export type SignInDataType = {
	email: string
	password: string
}

export type SignUpDataType = {
	email: string
	password: string
	username: string
}
