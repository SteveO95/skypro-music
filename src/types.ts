// Структура Трека
export type trackType = {
	dispatch(arg0: void): unknown;
	id: number;
	name: string;
	author: string;
	release_date: string;
	genre: string;
	duration_in_seconds: number;
	album: string;
	logo: null | string;
	track_file: string;
	stared_user: Array<userType>;
};

type userType = {
	id: number;
	username: string;
	first_name: string;
	last_name: string;
	email: string;
};
export type getTrackResponse = {
	error: string | null;
	data: trackType[] | null;
};
