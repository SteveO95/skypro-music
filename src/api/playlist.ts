import { fetchWithAuth } from "@/utils/fetchWithAuth";
const BASE_URL = "https://webdev-music-003b5b991590.herokuapp.com/";
const USER_URL = `${BASE_URL}user/`;
const TRACKS_URL = `${BASE_URL}catalog/track/`;
const SELECT_URL = `${BASE_URL}catalog/selection/`;

export async function getPlaylist() {
  const res = await fetch(TRACKS_URL + `all/`);
  if (!res.ok) {
    throw new Error(res.statusText);
  }
  const data = await res.json();
  return data.data;
}

export async function getSelectionPlaylist(id: string) {
  const res = await fetch(SELECT_URL + id);
  if (!res.ok) {
    throw new Error(res.statusText);
  }

  const json = await res.json();

  if (!json.data || !json.data.items) {
    console.error("Данные:", json);
    throw new Error(
      "Ошибка данных: Плейлист не найден или отсутствуют элементы"
    );
  }

  return json.data.items;
}

export async function likeTrack({
  trackId,
  access,
  refresh,
}: {
  trackId: number;
  access: string;
  refresh: string;
}) {
  const res = await fetchWithAuth(
    TRACKS_URL + `${trackId}/favorite/`,
    {
      method: "POST",
      headers: {
        Authorization: `Bearer ${access}`,
      },
    },
    refresh
  );

  if (!res.ok) {
    const errorData = await res.json();
    throw new Error(
      `Ошибка ${res.status}: ${errorData.message || "Неизвестная ошибка"}`
    );
  }
  return res.json();
}

export async function disLikeTrack({
  trackId,
  access,
  refresh,
}: {
  trackId: number;
  access: string;
  refresh: string;
}) {
  const res = await fetchWithAuth(
    TRACKS_URL + `${trackId}/favorite/`,
    {
      method: "DELETE",
      headers: {
        Authorization: `Bearer ${access}`,
      },
    },
    refresh
  );
  if (!res.ok) {
    const errorData = await res.json();
    throw new Error(
      `Ошибка ${res.status}: ${errorData.message || "Неизвестная ошибка"}`
    );
  }
  return res.json();
}

export async function fetchFavoriteTracks({
  access,
  refresh,
}: {
  access: string;
  refresh: string;
}) {
  const res = await fetchWithAuth(
    TRACKS_URL + `favorite/all/`,
    {
      method: "GET",
      headers: {
        Authorization: `Bearer ${access}`,
      },
    },
    refresh
  );
  const data = await res.json();
  return data.data;
}

export async function signInUser({
  email,
  password,
}: {
  email: string;
  password: string;
}) {
  const response = await fetch(USER_URL + `login/`, {
    method: "POST",
    body: JSON.stringify({ email, password }),
    headers: {
      "Content-Type": "application/json",
    },
  });

  const json = await response.json();

  if (response.status === 401) {
    throw new Error("Пользователь с таким email или паролем не найден");
  }
  if (response.status === 500) {
    throw new Error("Сервер сломался. Попробуйте позже");
  }
  if (response.ok) {
    return json;
  } else {
    throw new Error("Email или пароль неверен");
  }
}

export async function signUpUser({
  email,
  password,
  username,
}: {
  email: string;
  password: string;
  username: string;
}) {
  const response = await fetch(USER_URL + `signup/`, {
    method: "POST",
    body: JSON.stringify({ email, password, username }),
    headers: {
      "Content-Type": "application/json",
    },
  });

  const json = await response.json();

  if (response.status === 403) {
    throw new Error("Введенный Email уже занят");
  }
  if (response.status === 412) {
    throw new Error("Пароль должен быть больше 6 символов");
  }
  if (response.status === 400) {
    throw new Error(
      "Только цифры и латинские буквы (или спец символы: @/./+/-/_)"
    );
  }
  if (response.status === 500) {
    throw new Error("Сервер сломался. Попробуйте позже");
  }
  if (response.ok) {
    return json;
  }
}

export async function getTokens({
  email,
  password,
}: {
  email: string;
  password: string;
}) {
  const response = await fetch(USER_URL + `token/`, {
    method: "POST",
    body: JSON.stringify({ email, password }),
    headers: {
      "Content-Type": "application/json",
    },
  });

  const json = await response.json();

  if (!response.ok) {
    throw new Error(json.detail);
  }

  return json;
}

export async function refreshToken(refresh: string) {
  const res = await fetch(USER_URL + `token/refresh/`, {
    method: "POST",
    body: JSON.stringify({
      refresh,
    }),
    headers: {
      "Content-Type": "application/json",
    },
  });
  if (!res.ok) {
    throw new Error("Не удалось обновить токен");
  }
  return res.json();
}
