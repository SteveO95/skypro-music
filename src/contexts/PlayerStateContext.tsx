"use client";

import { createContext, ReactNode, useContext, useState } from "react";

type PlayerStateContextProps = {
  isPlaying: boolean;
  setIsPlaying: React.Dispatch<React.SetStateAction<boolean>>;
};

const PlayerStateContext = createContext<PlayerStateContextProps | undefined>(
  undefined
);

type PlayerStateProviderProps = {
  children: ReactNode;
};

export function PlayerStateProvider({ children }: PlayerStateProviderProps) {
  const [isPlaying, setIsPlaying] = useState<boolean>(false);

  return (
    <PlayerStateContext.Provider value={{ isPlaying, setIsPlaying }}>
      {children}
    </PlayerStateContext.Provider>
  );
}

export function usePlayerState() {
  const context = useContext(PlayerStateContext);
  if (context === undefined) {
    throw new Error("usePlayerState должен использоваться внутри провайдера");
  }
  return context;
}
