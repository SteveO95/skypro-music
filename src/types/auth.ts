export type AuthType = {
  id: string | undefined;
  username: string | undefined;
  email: string | undefined;
  accessToken: string | undefined;
  refreshToken: string | undefined;
};
