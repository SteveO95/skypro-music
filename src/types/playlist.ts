import { UserType } from './user';

export type PlaylistType = {
	_id(_id: any): { isLiked: any; handleLike: any };
	id: number;
	name: string;
	author: string;
	release_date: string;
	genre: string;
	duration_in_seconds: number;
	album: string;
	logo: string | null;
	track_file: string;
	stared_user: UserType[];
};
