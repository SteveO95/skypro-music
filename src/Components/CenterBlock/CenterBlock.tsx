import { getPlaylist } from '../../api/playlist';
import Filter from '../../components/Filter/Filter';
import Playlist from '../../components/Playlist/Playlist';
import Search from '../../components/Search/Search';
import { PlaylistType } from '../../types/playlist';
import styles from './CenterBlock.module.css';

const CenterBlock = async () => {
	let tracks: PlaylistType[] = [];
	let error = '';
	try {
		tracks = await getPlaylist();
	} catch (err: unknown) {
		error =
			err instanceof Error
				? 'Ошибка при загрузки треков ' + err.message
				: 'Неизвестная ошибка';
	}
	return (
		<div className={styles.mainCenterblock}>
			<Search />
			<h2 className={styles.centerblockHeader}>Треки</h2>
			<Filter tracks={tracks} />
			{error && <div className={styles.error}>{error}</div>}
			<Playlist tracks={tracks} />
		</div>
	);
};

export default CenterBlock;
