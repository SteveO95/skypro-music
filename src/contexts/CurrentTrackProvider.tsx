"use client";

import { PlaylistType } from "@/types/playlist";
import {
  createContext,
  Dispatch,
  ReactNode,
  SetStateAction,
  useContext,
  useState,
} from "react";

type CurrentTrackContextValue = {
  currentTrack: PlaylistType | null;
  setCurrentTrack: Dispatch<SetStateAction<PlaylistType | null>>;
};

const CurrentTrackContext = createContext<CurrentTrackContextValue | undefined>(
  undefined
);

type CurrentTrackProviderProps = {
  children: ReactNode;
};

export function CurrentTrackProvider({ children }: CurrentTrackProviderProps) {
  const [currentTrack, setCurrentTrack] = useState<PlaylistType | null>(null);

  return (
    <CurrentTrackContext.Provider value={{ currentTrack, setCurrentTrack }}>
      {children}
    </CurrentTrackContext.Provider>
  );
}

export function useCurrentTrack() {
  const context = useContext(CurrentTrackContext);
  if (context === undefined) {
    throw new Error("useCurrentTrack должен использоваться внутри провайдера");
  }
  return context;
}
