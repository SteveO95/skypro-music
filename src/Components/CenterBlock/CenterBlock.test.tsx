import React from "react";
import renderer from "react-test-renderer";
import CenterBlock from "./CenterBlock";
import { PlaylistType } from "@/types/playlist";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { PlayerStateContext } from "@/contexts/PlayerStateContext";

const mockStore = configureStore();
const initialState = {
  user: {
    tokens: {
      access: "mockAccessToken",
      refresh: "mockRefreshToken",
    },
  },
  playlist: {
    likedPlaylist: [],
    tracks: [],
  },
  filters: {
    keyword: "",
    genres: [],
    authors: [],
    sortOrder: "По умолчанию",
  },
};
const store = mockStore(initialState);

const mockPlayerState = {
  isPlaying: false,
  setIsPlaying: jest.fn(),
};

it("корректный рендеринг треков и заголовков", () => {
  const mockTracks: PlaylistType[] = [
    {
      _id: 1,
      name: "Test Track 1",
      author: "Test Artist 1",
      release_date: "2024-01-01",
      genre: "Pop",
      duration_in_seconds: 200,
      album: "Test Album 1",
      logo: "/path/to/logo1.png",
      track_file: "/path/to/track1.mp3",
      stared_user: [],
    },
    {
      _id: 2,
      name: "Test Track 2",
      author: "Test Artist 2",
      release_date: "2024-02-01",
      genre: "Rock",
      duration_in_seconds: 210,
      album: "Test Album 2",
      logo: "/path/to/logo2.png",
      track_file: "/path/to/track2.mp3",
      stared_user: [],
    },
  ];

  const tree = renderer
    .create(
      <Provider store={store}>
        <PlayerStateContext.Provider value={mockPlayerState}>
          <CenterBlock tracks={mockTracks} title="Test Title" />
        </PlayerStateContext.Provider>
      </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it("корректный рендеринг с ошибкой", () => {
  const mockTracks: PlaylistType[] = [];

  const tree = renderer
    .create(
      <Provider store={store}>
        <PlayerStateContext.Provider value={mockPlayerState}>
          <CenterBlock tracks={mockTracks} error="Произошла ошибка" />
        </PlayerStateContext.Provider>
      </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
