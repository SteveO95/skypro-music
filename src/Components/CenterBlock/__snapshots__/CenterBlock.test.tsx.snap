// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`корректный рендеринг с ошибкой 1`] = `
<div
  className="mainCenterblock"
>
  <div
    className="centerblockSearch"
  >
    <svg
      className="searchSvg"
    >
      <use
        xlinkHref="/img/icon/sprite.svg#icon-search"
      />
    </svg>
    <input
      className="searchText"
      name="search"
      onChange={[Function]}
      placeholder="Поиск"
      type="search"
    />
  </div>
  <h2
    className="centerblockHeader"
  >
    Треки
  </h2>
  <div
    className="centerblockFilter"
  >
    <div
      className="filterTitle"
    >
      Искать по:
    </div>
    <div
      className="filterWrapper"
    >
      <div
        className="filterButton"
        onClick={[Function]}
      >
        исполнителю
      </div>
    </div>
    <div
      className="filterWrapper"
    >
      <div
        className="filterButton"
        onClick={[Function]}
      >
        году выпуска
      </div>
    </div>
    <div
      className="filterWrapper"
    >
      <div
        className="filterButton"
        onClick={[Function]}
      >
        жанру
      </div>
    </div>
  </div>
  <div
    className="error"
  >
    Произошла ошибка
  </div>
  <div
    className="noTracks"
  >
    Треки не найдены
  </div>
</div>
`;

exports[`корректный рендеринг треков и заголовков 1`] = `
<div
  className="mainCenterblock"
>
  <div
    className="centerblockSearch"
  >
    <svg
      className="searchSvg"
    >
      <use
        xlinkHref="/img/icon/sprite.svg#icon-search"
      />
    </svg>
    <input
      className="searchText"
      name="search"
      onChange={[Function]}
      placeholder="Поиск"
      type="search"
    />
  </div>
  <h2
    className="centerblockHeader"
  >
    Test Title
  </h2>
  <div
    className="centerblockFilter"
  >
    <div
      className="filterTitle"
    >
      Искать по:
    </div>
    <div
      className="filterWrapper"
    >
      <div
        className="filterButton"
        onClick={[Function]}
      >
        исполнителю
      </div>
    </div>
    <div
      className="filterWrapper"
    >
      <div
        className="filterButton"
        onClick={[Function]}
      >
        году выпуска
      </div>
    </div>
    <div
      className="filterWrapper"
    >
      <div
        className="filterButton"
        onClick={[Function]}
      >
        жанру
      </div>
    </div>
  </div>
  <div
    className="centerblockContent"
  >
    <div
      className="contentTitle"
    >
      <div
        className="playlistTitleCol col01"
      >
        Трек
      </div>
      <div
        className="playlistTitleCol col02"
      >
        Исполнитель
      </div>
      <div
        className="playlistTitleCol col03"
      >
        Альбом
      </div>
      <div
        className="playlistTitleCol col04"
      >
        <svg
          className="playlistTitleSvg"
        >
          <use
            xlinkHref="/img/icon/sprite.svg#icon-watch"
          />
        </svg>
      </div>
    </div>
    <div
      className="contentPlaylist"
    >
      <div
        className="playlistItem"
      >
        <div
          className="playlistTrack"
        >
          <div
            className="trackTitle"
            onClick={[Function]}
          >
            <div
              className="trackTitleImage"
            >
              <svg
                className="trackTitleSvg"
              >
                <use
                  xlinkHref="/img/icon/sprite.svg#icon-note"
                />
              </svg>
            </div>
            <div
              className="track__title-text"
            >
              <span
                className="trackTitleLink"
              >
                Test Track 1
                <span
                  className="trackTitleSpan"
                />
              </span>
            </div>
          </div>
          <div
            className="trackAuthor"
          >
            <span
              className="trackAuthorLink"
            >
              Test Artist 1
            </span>
          </div>
          <div
            className="trackAlbum"
          >
            <span
              className="trackAlbumLink"
            >
              Test Album 1
            </span>
          </div>
          <div
            onClick={[Function]}
          >
            <svg
              className="trackTimeSvg"
            >
              <use
                xlinkHref="/img/icon/sprite.svg#icon-dislike"
              />
            </svg>
            <span
              className="trackTimeText"
            >
              03:20
            </span>
          </div>
        </div>
      </div>
      <div
        className="playlistItem"
      >
        <div
          className="playlistTrack"
        >
          <div
            className="trackTitle"
            onClick={[Function]}
          >
            <div
              className="trackTitleImage"
            >
              <svg
                className="trackTitleSvg"
              >
                <use
                  xlinkHref="/img/icon/sprite.svg#icon-note"
                />
              </svg>
            </div>
            <div
              className="track__title-text"
            >
              <span
                className="trackTitleLink"
              >
                Test Track 2
                <span
                  className="trackTitleSpan"
                />
              </span>
            </div>
          </div>
          <div
            className="trackAuthor"
          >
            <span
              className="trackAuthorLink"
            >
              Test Artist 2
            </span>
          </div>
          <div
            className="trackAlbum"
          >
            <span
              className="trackAlbumLink"
            >
              Test Album 2
            </span>
          </div>
          <div
            onClick={[Function]}
          >
            <svg
              className="trackTimeSvg"
            >
              <use
                xlinkHref="/img/icon/sprite.svg#icon-dislike"
              />
            </svg>
            <span
              className="trackTimeText"
            >
              03:30
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
